### PR TITLE
Evaluation-time builds

### LINKED ISSUE #88: IFD should dispatch builds via Hercules CI
This become especially painful when mixing systems, as it currently can't dispatch IFD builds to darwin/linux.

Resulting into 

```
Exception: a 'x86_64-darwin' with features {} is required to build '/nix/store/vpmgblfzp2bi998f9rfggra5hnga7kjl-cabal2nix-hercules-ci-agent.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, kvm, nixos-test}; type: nix::Error
```

### PR DESCRIPTION
**Pull request by roberth**
Fixes #88

### CODE REVIEW

**Comment by domenkozar** [comment]
Looks good - I think we should implement the fetch early optimization as it will be the common path on most of the evaluations

File: hercules-ci-agent/hercules-ci-agent/Hercules/Agent/Evaluate.hs [hunk 1]
```diff
@@ -323,6 +339,27 @@ runEvalProcess projectDir file autoArguments nixPath emit = do
...
                     , Message.message = e
                     }
+            Event.Build drv outputName -> do
+              status <- unlift $ withNamedContext "derivation" drv $ do
+                currentIndex <- liftIO $ atomicModifyIORef buildRequiredIndex (\i -> (i + 1, i))
+                liftIO $ emit $ EvaluateEvent.BuildRequired BuildRequired.BuildRequired
+                    { BuildRequired.derivationPath = drv
+                    , BuildRequired.index = currentIndex
+                    , BuildRequired.outputName = outputName
+                    }
```
Comment (roberth) [inline hercules-ci-agent/hercules-ci-agent/Hercules/Agent/Evaluate.hs]: TODO: push `DerivationInfo`s

**Comment by domenkozar** [comment]
- [ ] bump to cachix master to get bulk query optimization

**Comment by roberth** [comment]
Ready for review

**Comment by domenkozar** [comment]
Opened https://github.com/hercules-ci/hercules-ci-agent/issues/128 https://github.com/hercules-ci/hercules-ci-agent/issues/129

**Pull request merged**